### ISSUE TITLE
Bug: Wrong argument type Integer (expected Proc)

### ISSUE THREAD

**Issue by marceloboeira** [DatabaseCleaner/database_cleaner#466]
Running my tests end up on this:
```
 Error: DatabaseCleaner.clean
 TypeError:
 wrong argument type Integer (expected Proc)
 # ./spec/support/auto/database_cleaner.rb:16:in `block (2 levels) in <top (required)>'
```

Ruby 2.4.0

**UPDATE** it does not happen with Ruby 2.3.1.

Gemfile.lock:
```
GIT
  remote: https://github.com/kule/podcast_reader
  revision: 82328f88e5715b1f60f40e40c589af65287e2393
  specs:
    podcast_reader (0.1.0)

GIT
  remote: https://github.com/marceloboeira/mongoid-uuid
  revision: be0cfa220d72019a06f292cab071ca8a5bde982f
  specs:
    mongoid-uuid (1.1.0)
      mongoid (>= 3.0, < 7)
      rake
      uuid (~> 2.3)

GEM
  remote: https://rubygems.org/
  specs:
    actioncable (5.0.1)
      actionpack (= 5.0.1)
      nio4r (~> 1.2)
      websocket-driver (~> 0.6.1)
    actionmailer (5.0.1)
      actionpack (= 5.0.1)
      actionview (= 5.0.1)
      activejob (= 5.0.1)
      mail (~> 2.5, >= 2.5.4)
      rails-dom-testing (~> 2.0)
    actionpack (5.0.1)
      actionview (= 5.0.1)
      activesupport (= 5.0.1)
      rack (~> 2.0)
      rack-test (~> 0.6.3)
      rails-dom-testing (~> 2.0)
      rails-html-sanitizer (~> 1.0, >= 1.0.2)
    actionview (5.0.1)
      activesupport (= 5.0.1)
      builder (~> 3.1)
      erubis (~> 2.7.0)
      rails-dom-testing (~> 2.0)
      rails-html-sanitizer (~> 1.0, >= 1.0.2)
    activejob (5.0.1)
      activesupport (= 5.0.1)
      globalid (>= 0.3.6)
    activemodel (5.0.1)
      activesupport (= 5.0.1)
    activerecord (5.0.1)
      activemodel (= 5.0.1)
      activesupport (= 5.0.1)
      arel (~> 7.0)
    activesupport (5.0.1)
      concurrent-ruby (~> 1.0, >= 1.0.2)
      i18n (~> 0.7)
      minitest (~> 5.1)
      tzinfo (~> 1.1)
    addressable (2.5.0)
      public_suffix (~> 2.0, >= 2.0.2)
    arel (7.1.4)
    bson (4.2.1)
    builder (3.2.2)
    byebug (9.0.6)
    capybara (2.11.0)
      addressable
      mime-types (>= 1.16)
      nokogiri (>= 1.3.3)
      rack (>= 1.0.0)
      rack-test (>= 0.5.4)
      xpath (~> 2.0)
    coderay (1.1.1)
    coffee-rails (4.2.1)
      coffee-script (>= 2.2.0)
      railties (>= 4.0.0, < 5.2.x)
    coffee-script (2.4.1)
      coffee-script-source
      execjs
    coffee-script-source (1.12.2)
    concurrent-ruby (1.0.4)
    database_cleaner (1.5.3)
    debug_inspector (0.0.2)
    diff-lcs (1.2.5)
    erubis (2.7.0)
    execjs (2.7.0)
    ffi (1.9.14)
    font-awesome-rails (4.7.0.1)
      railties (>= 3.2, < 5.1)
    formatador (0.2.5)
    functional_operations (2.3.0)
    globalid (0.3.7)
      activesupport (>= 4.1.0)
    guard (2.14.0)
      formatador (>= 0.2.4)
      listen (>= 2.7, < 4.0)
      lumberjack (~> 1.0)
      nenv (~> 0.1)
      notiffany (~> 0.0)
      pry (>= 0.9.12)
      shellany (~> 0.0)
      thor (>= 0.18.1)
    guard-bundler (2.1.0)
      bundler (~> 1.0)
      guard (~> 2.2)
      guard-compat (~> 1.1)
    guard-compat (1.2.1)
    guard-rails (0.8.0)
      guard (~> 2.11)
      guard-compat (~> 1.0)
    guard-rspec (4.7.3)
      guard (~> 2.1)
      guard-compat (~> 1.1)
      rspec (>= 2.99.0, < 4.0)
    haml (4.0.7)
      tilt
    i18n (0.7.0)
    jquery-rails (4.2.2)
      rails-dom-testing (>= 1, < 3)
      railties (>= 4.2.0)
      thor (>= 0.14, < 2.0)
    jquery-ui-rails (5.0.5)
      railties (>= 3.2.16)
    kaminari (0.17.0)
      actionpack (>= 3.0.0)
      activesupport (>= 3.0.0)
    listen (3.0.8)
      rb-fsevent (~> 0.9, >= 0.9.4)
      rb-inotify (~> 0.9, >= 0.9.7)
    loofah (2.0.3)
      nokogiri (>= 1.5.9)
    lumberjack (1.0.10)
    macaddr (1.7.1)
      systemu (~> 2.6.2)
    mail (2.6.4)
      mime-types (>= 1.16, < 4)
    method_source (0.8.2)
    mime-types (3.1)
      mime-types-data (~> 3.2015)
    mime-types-data (3.2016.0521)
    mini_portile2 (2.1.0)
    minitest (5.10.1)
    mongo (2.4.1)
      bson (>= 4.2.1, < 5.0.0)
    mongoid (6.0.3)
      activemodel (~> 5.0)
      mongo (~> 2.3)
    nenv (0.3.0)
    nested_form (0.3.2)
    nio4r (1.2.1)
    nokogiri (1.7.0)
      mini_portile2 (~> 2.1.0)
    notiffany (0.1.1)
      nenv (~> 0.1)
      shellany (~> 0.0)
    pry (0.10.4)
      coderay (~> 1.1.0)
      method_source (~> 0.8.1)
      slop (~> 3.4)
    public_suffix (2.0.4)
    puma (3.6.2)
    rack (2.0.1)
    rack-pjax (1.0.0)
      nokogiri (~> 1.5)
      rack (>= 1.1)
    rack-test (0.6.3)
      rack (>= 1.0)
    rails (5.0.1)
      actioncable (= 5.0.1)
      actionmailer (= 5.0.1)
      actionpack (= 5.0.1)
      actionview (= 5.0.1)
      activejob (= 5.0.1)
      activemodel (= 5.0.1)
      activerecord (= 5.0.1)
      activesupport (= 5.0.1)
      bundler (>= 1.3.0, < 2.0)
      railties (= 5.0.1)
      sprockets-rails (>= 2.0.0)
    rails-dom-testing (2.0.2)
      activesupport (>= 4.2.0, < 6.0)
      nokogiri (~> 1.6)
    rails-html-sanitizer (1.0.3)
      loofah (~> 2.0)
    rails_admin (1.1.1)
      builder (~> 3.1)
      coffee-rails (~> 4.0)
      font-awesome-rails (>= 3.0, < 5)
      haml (~> 4.0)
      jquery-rails (>= 3.0, < 5)
      jquery-ui-rails (~> 5.0)
      kaminari (~> 0.14)
      nested_form (~> 0.3)
      rack-pjax (>= 0.7)
      rails (>= 4.0, < 6)
      remotipart (~> 1.3)
      sass-rails (>= 4.0, < 6)
    railties (5.0.1)
      actionpack (= 5.0.1)
      activesupport (= 5.0.1)
      method_source
      rake (>= 0.8.7)
      thor (>= 0.18.1, < 2.0)
    rake (12.0.0)
    rb-fsevent (0.9.8)
    rb-inotify (0.9.7)
      ffi (>= 0.5.0)
    remotipart (1.3.1)
    rspec (3.5.0)
      rspec-core (~> 3.5.0)
      rspec-expectations (~> 3.5.0)
      rspec-mocks (~> 3.5.0)
    rspec-core (3.5.4)
      rspec-support (~> 3.5.0)
    rspec-expectations (3.5.0)
      diff-lcs (>= 1.2.0, < 2.0)
      rspec-support (~> 3.5.0)
    rspec-mocks (3.5.0)
      diff-lcs (>= 1.2.0, < 2.0)
      rspec-support (~> 3.5.0)
    rspec-rails (3.5.2)
      actionpack (>= 3.0)
      activesupport (>= 3.0)
      railties (>= 3.0)
      rspec-core (~> 3.5.0)
      rspec-expectations (~> 3.5.0)
      rspec-mocks (~> 3.5.0)
      rspec-support (~> 3.5.0)
    rspec-support (3.5.0)
    sass (3.4.23)
    sass-rails (5.0.6)
      railties (>= 4.0.0, < 6)
      sass (~> 3.1)
      sprockets (>= 2.8, < 4.0)
      sprockets-rails (>= 2.0, < 4.0)
      tilt (>= 1.1, < 3)
    shellany (0.0.1)
    slop (3.6.0)
    sprockets (3.7.1)
      concurrent-ruby (~> 1.0)
      rack (> 1, < 3)
    sprockets-rails (3.2.0)
      actionpack (>= 4.0)
      activesupport (>= 4.0)
      sprockets (>= 3.0.0)
    systemu (2.6.5)
    thor (0.19.4)
    thread_safe (0.3.5)
    tilt (2.0.5)
    tzinfo (1.2.2)
  thread_safe (~> 0.1)
    uglifier (3.0.4)
      execjs (>= 0.3.0, < 3)
    uuid (2.3.8)
      macaddr (~> 1.0)
    web-console (3.4.0)
      actionview (>= 5.0)
      activemodel (>= 5.0)
      debug_inspector
      railties (>= 5.0)
    websocket-driver (0.6.4)
      websocket-extensions (>= 0.1.0)
    websocket-extensions (0.1.2)
    xpath (2.0.0)
      nokogiri (~> 1.3)

PLATFORMS
  ruby

DEPENDENCIES
  byebug
  capybara
  database_cleaner
  functional_operations
  guard-bundler
  guard-rails
  guard-rspec
  listen (~> 3.0.5)
  mongoid (~> 6.0.0)
  mongoid-uuid!
  nokogiri
  podcast_reader!
  puma (~> 3.0)
  rails (~> 5.0.1)
  rails_admin (~> 1.0)
  rspec-rails (~> 3.5)
  sass-rails (~> 5.0)
  uglifier (>= 1.3.0)
  web-console (>= 3.3.0)

BUNDLED WITH
   1.13.7
```

**Comment by dwbutler** [comment]
Same. Using truncation strategy with Mongoid 6, Database Cleaner 1.5.3, Ruby 2.4.0.

```ruby
Failure/Error: DatabaseCleaner[:mongoid].clean
          
TypeError:
  wrong argument type Integer (expected Proc)
# ./gemfiles/vendor/bundle/ruby/2.4.0/gems/mongo-2.4.1/lib/mongo/server/connection.rb:195:in `block in write'
# ./gemfiles/vendor/bundle/ruby/2.4.0/gems/mongo-2.4.1/lib/mongo/server/connection.rb:192:in `each'
# ./gemfiles/vendor/bundle/ruby/2.4.0/gems/mongo-2.4.1/lib/mongo/server/connection.rb:192:in `write'
# ./gemfiles/vendor/bundle/ruby/2.4.0/gems/mongo-2.4.1/lib/mongo/server/connection.rb:162:in `deliver'
# ./gemfiles/vendor/bundle/ruby/2.4.0/gems/mongo-2.4.1/lib/mongo/server/connection.rb:108:in `block in dispatch'
# ./gemfiles/vendor/bundle/ruby/2.4.0/gems/mongo-2.4.1/lib/mongo/monitoring/publishable.rb:47:in `publish_command'
# ./gemfiles/vendor/bundle/ruby/2.4.0/gems/mongo-2.4.1/lib/mongo/server/connection.rb:107:in `dispatch'
# ./gemfiles/vendor/bundle/ruby/2.4.0/gems/mongo-2.4.1/lib/mongo/operation/write/gle.rb:31:in `block in execute_message'
# ./gemfiles/vendor/bundle/ruby/2.4.0/gems/mongo-2.4.1/lib/mongo/server/connection_pool.rb:107:in `with_connection'
# ./gemfiles/vendor/bundle/ruby/2.4.0/gems/mongo-2.4.1/lib/mongo/server.rb:242:in `with_connection'
# ./gemfiles/vendor/bundle/ruby/2.4.0/gems/mongo-2.4.1/lib/mongo/operation/write/gle.rb:28:in `execute_message'
# ./gemfiles/vendor/bundle/ruby/2.4.0/gems/mongo-2.4.1/lib/mongo/operation/write/write_command_enabled.rb:39:in `execute'
# ./gemfiles/vendor/bundle/ruby/2.4.0/gems/mongo-2.4.1/lib/mongo/collection/view/writable.rb:224:in `block in remove'
# ./gemfiles/vendor/bundle/ruby/2.4.0/gems/mongo-2.4.1/lib/mongo/retryable.rb:112:in `write_with_retry'
# ./gemfiles/vendor/bundle/ruby/2.4.0/gems/mongo-2.4.1/lib/mongo/collection/view/writable.rb:218:in `remove'
# ./gemfiles/vendor/bundle/ruby/2.4.0/gems/mongo-2.4.1/lib/mongo/collection/view/writable.rb:136:in `delete_many'
# ./gemfiles/vendor/bundle/ruby/2.4.0/gems/mongoid-6.0.0/lib/mongoid/query_cache.rb:168:in `delete_many_with_clear_cache'
# ./gemfiles/vendor/bundle/ruby/2.4.0/gems/database_cleaner-1.5.3/lib/database_cleaner/mongo2/truncation_mixin.rb:9:in `block in clean'
# ./gemfiles/vendor/bundle/ruby/2.4.0/gems/database_cleaner-1.5.3/lib/database_cleaner/mongo2/truncation_mixin.rb:9:in `each'
# ./gemfiles/vendor/bundle/ruby/2.4.0/gems/database_cleaner-1.5.3/lib/database_cleaner/mongo2/truncation_mixin.rb:9:in `clean'
# ./gemfiles/vendor/bundle/ruby/2.4.0/gems/database_cleaner-1.5.3/lib/database_cleaner/base.rb:92:in `clean'
# ./spec/mongoid_spec.rb:13:in `block (2 levels) in <top (required)>'
```

**Comment by dwbutler** [comment]
On closer inspection, this is an issue with the `mongo` gem, not Database Cleaner. See https://github.com/mongodb/mongo-ruby-driver/pull/847.

**Comment by marceloboeira** [comment]
@dwbutler thanks! Do we close this one?

**Comment by dwbutler** [comment]
Yes. There isn't anything for Database Cleaner to fix here. We just have to wait until a new version of the mongo driver is released that is compatible with Ruby 2.4.0.

**Comment by twalpole** [comment]
FYI - This isn't an issue with mongo, it's an issue with MRI Ruby 2.4.0 - Anything using delegators via forwardable.rb is potentially broken.  https://bugs.ruby-lang.org/issues/13107   Fix will come in Ruby 2.4.1

**Comment by marceloboeira** [comment]
@twalpole 👍

**Comment by ronak21691** [comment]
I am facing the same issue, any idea when ruby 2.4.1 will be released or is there any work around for the time being.

**Comment by twalpole** [comment]
@ronak21691  There is no published timetable for 2.4.1 - workaround is to revert to 2.3.3

**Comment by ronak21691** [comment]
Thanks, working fine now.

**Comment by ncreuschling** [comment]
Ruby v2.4.1 just got released. Can somebody confirm if it works now.

**Comment by dwbutler** [comment]
I can confirm that this issue is fixed under Ruby 2.4.1.
